src/BuildScriptGenerator/Node/NodeBashBuildSnippet.sh.tpl

```diff
@@ -1,44 +1,104 @@
 then
 	echo
-    echo "Configuring Yarn cache folder ..."
+    echo "Configuring Yarn cache folder..."
     yarn config set cache-folder $YARN_CACHE_DIR
 fi
 
+allNodeModulesDirName=__oryx_all_node_modules
+if [ -d "$SOURCE_DIR"/$allNodeModulesDirName ]
+then
+	# Clean-up any earlier build artifacts (which contain only prod dependencies)
+	rm -rf node_modules
+	rm -f node_modules.zip
+
+	# Rename the folder which has all the node modules to reuse for later builds to improve perf
+	mv "$SOURCE_DIR"/$allNodeModulesDirName "$SOURCE_DIR"/node_modules
+fi
+
+# Directory where production dependencies are installed
+prodDependencies="/tmp/prod-dependencies"
+
+hasProductionOnlyDependencies="{{ HasProductionOnlyDependencies }}"
+if [ "$SOURCE_DIR" != "$DESTINATION_DIR" ] && [ "$hasProductionOnlyDependencies" == "true" ]
+then
+	mkdir -p "$prodDependencies"
+	cd "$prodDependencies"
+
+	if [ -f "$SOURCE_DIR/package.json" ]; then
+		cp -f "$SOURCE_DIR/package.json" .
+	fi
+
+	if [ -f "$SOURCE_DIR/package-lock.json" ]; then
+		cp -f "$SOURCE_DIR/package-lock.json" .
+	fi
+
+	if [ -f "$SOURCE_DIR/yarn.lock" ]; then
+		cp -f "$SOURCE_DIR/yarn.lock" .
+	fi
+
+	echo "Installing production dependencies in '$prodDependencies'..."
+	echo
+	echo "Running '{{ ProductionOnlyPackageInstallCommand }}'..."
+	echo
+	{{ ProductionOnlyPackageInstallCommand }}
+
+	echo
+	echo "Copying production dependencies from '$prodDependencies' to '$SOURCE_DIR'..."
+	rsync -rtE --links node_modules "$SOURCE_DIR"
+fi
+
+# Install all dependencies
+cd "$SOURCE_DIR"
+echo "Installing dependencies..."
 echo
-echo Installing packages ...
-echo
-echo "Running '{{ PackageInstallCommand }}' ..."
+echo "Running '{{ PackageInstallCommand }}'..."
 echo
 {{ PackageInstallCommand }}
 
 {{ if NpmRunBuildCommand | IsNotBlank }}
 echo
-echo "Running '{{ NpmRunBuildCommand }}' ..."
+echo "Running '{{ NpmRunBuildCommand }}'..."
 echo
 {{ NpmRunBuildCommand }}
 {{ end }}
 
 {{ if NpmRunBuildAzureCommand | IsNotBlank }}
 echo
-echo "Running '{{ NpmRunBuildAzureCommand }}' ..."
+echo "Running '{{ NpmRunBuildAzureCommand }}'..."
 echo
 {{ NpmRunBuildAzureCommand }}
 {{ end }}
 
-{{ if ZipNodeModulesDir }}
-# If source and destination are the same, then we need not zip the contents again, but if they are different
-# then we want to copy the zipped node modules to destination directory as this directory could be in a shared volume.
 if [ "$SOURCE_DIR" != "$DESTINATION_DIR" ]
 then
-	echo
-	echo Zipping 'node_modules' folder ...
-	rm -f "node_modules.zip"
-	tar -zcf node_modules.zip node_modules
+	if [ "$hasProductionOnlyDependencies" == "true" ]
+	then
+		# Rename folder having all node_modules as we want only prod dependencies
+		# to be synced with destination directory
+		mv node_modules __oryx_all_node_modules
+		
+		echo
+		echo "Copying production dependencies from '$prodDependencies' to '$SOURCE_DIR'..."
+		echo
+		cd "$prodDependencies"
+		{{ if ZipNodeModulesDir }}
+			echo
```

Comment:
I know this indentation shows up in the generated script but the script is not easily readable without it.